tdx/mrtd_from_ovmf.go

```diff
@@ -18,38 +18,61 @@ import (
 	"fmt"
 
 	"github.com/google/gce-tcb-verifier/ovmf"
+	"github.com/google/gce-tcb-verifier/ovmf/abi"
 )
 
 const (
-	mib = 0x100000
-	gib = 0x40000000
+	mib           = 0x100000
+	gib           = 0x40000000
+	mmioHoleStart = 3 * gib
+	mmioHoleEnd   = 4 * gib
 )
 
-func regionsForGB(sizeInGB int) []ovmf.GuestPhysicalRegion {
-	return []ovmf.GuestPhysicalRegion{
-		{Start: 0, Length: 3 * gib},                        // Lower 3GiB
-		{Start: 4*gib - 2*mib, Length: 2 * mib},            // Top 2MiB under 4GiB for TDVF
-		{Start: 4 * gib, Length: uint64(sizeInGB-3) * gib}, // Rest of memory
+type numaShape struct {
+	size           int
+	nodes          int
+	maxSizePerNode int
+}
+
+func regionsForShape(shape numaShape) []ovmf.GuestPhysicalRegion {
+	regions := []ovmf.GuestPhysicalRegion{
+		{Start: 0, Length: mmioHoleStart},       // Lower 3GiB
+		{Start: 4*gib - 2*mib, Length: 2 * mib}, // Top 2MiB under 4GiB for TDVF
+	}
+	Start := abi.EFIPhysicalAddress(mmioHoleEnd)
+	taken := abi.EFIPhysicalAddress(mmioHoleStart)
+	Size := uint64(abi.EFIPhysicalAddress(shape.size*gib) - taken)
+	MaxNodeSize := uint64(shape.maxSizePerNode * gib)
+	for node := 0; node < shape.nodes; node++ {
+		length := MaxNodeSize - uint64(taken)
+		if Size < length {
+			length = Size
+		}
+		regions = append(regions, ovmf.GuestPhysicalRegion{Start: Start, Length: length})
+		Start += abi.EFIPhysicalAddress(length)
+		Size -= length
+		taken = 0
 	}
+	return regions
 }
 
 // https://cloud.google.com/compute/docs/general-purpose-machines#c3_machine_types
-var shapeRAMGib = map[string]int{
-	"c3-standard-4":   16,
-	"c3-standard-8":   32,
-	"c3-standard-22":  88,
-	"c3-standard-44":  176,
-	"c3-standard-88":  352,
-	"c3-standard-176": 704,
+var shapeDesc = map[string]numaShape{
+	"c3-standard-4":   {size: 16, nodes: 1, maxSizePerNode: 176},
+	"c3-standard-8":   {size: 32, nodes: 1, maxSizePerNode: 176},
+	"c3-standard-22":  {size: 88, nodes: 1, maxSizePerNode: 176},
+	"c3-standard-44":  {size: 176, nodes: 1, maxSizePerNode: 176},
+	"c3-standard-88":  {size: 352, nodes: 2, maxSizePerNode: 176},
+	"c3-standard-176": {size: 704, nodes: 4, maxSizePerNode: 176},
 }
 
 var shapeBanks = map[string][]ovmf.GuestPhysicalRegion{
```

Comment:
Seems like shapeBanks could be calculated automatically rather than respecifying each shape. Could use a call to map or just replace uses of shapeBanks with calls to regionsForShape

Reply:
Done.